crates/red_knot/src/cache.rs

```diff
@@ -7,15 +7,19 @@ use dashmap::mapref::entry::Entry;
 use crate::FxDashMap;
 
 /// Simple key value cache that locks on a per-key level.
-pub struct KeyValueCache<K, V> {
+pub struct KeyValueCache<K, V, S = DefaultStatisticsRecorder>
+where
+    S: StatisticsRecorder,
+{
     map: FxDashMap<K, V>,
-    statistics: CacheStatistics,
+    statistics: S,
 }
```

Comment:
I like the trait, but making `KeyValueCache` generic over `S` feels heavyweight because it increases the complexity of using the type (yes, it defaults to a type, but you still get scared away by three generics when looking at the type).

Reply:
Makes sense. Is there a way of using a trait here without making `KeyValueCache` generic? If not, I'm not wedded to this approach -- happy to do something else with this, or not do anything at all!

Reply:
I suppose I could do `statistics: Box<dyn StatisticsRecorder>`?

Reply:
You could, but that requires an allocation and adds an overhead to cache lookups (even when it's the release statistics). 

I think we should not over-engineer this. We might even go with always having statistics. This is just some prototype code.

Reply:
Makes sense. Should I just close this for now, then? :-)

Reply:
I like the `trait`. I would just remove the type param.

Reply:
Ohh, I think I see what you mean. Sorry for being dense.